src/node.cc

```diff
@@ -977,82 +1008,78 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
   return 0;
 }
 
-InitializationResult InitializeOncePerProcess(int argc, char** argv) {
-  return InitializeOncePerProcess(argc, argv, kDefaultInitialization);
-}
+std::unique_ptr<InitializationResult> InitializeOncePerProcess(
```

Comment:
I'm not sure I fully understand why this is now wrapping the InitializationResult in a unique_ptr now. Is it because of:
> The return value of the function is made an abstract class,
> rather than a struct, for easier API/ABI stability.

? Does it mean that we will be able to change the `InitializationResult` part into something else in the future if the need arises, without breaking ABI compatibility?

Reply:
> Is it because of:
> 
> > The return value of the function is made an abstract class,
> > rather than a struct, for easier API/ABI stability.
> 
> ? 

Exactly :+1:

> Does it mean that we will be able to change the `InitializationResult` part into something else in the future if the need arises, without breaking ABI compatibility?

Well, there are always limitations, but we could add new fields or change the internal representation of existing fields without breaking ABI compatibility. (There’s also more extensive solutions to this, of course, e.g. how the `CommonEnvironmentSetup` class avoids virtual functions and instead uses an internal struct to define its properties. I think that would be a bit much here but I’m happy to move to that model if that’s preferred.)

Reply:
Are there other places in the embedder API where we don't go as far as CommonEnvironmentSetup? If we want to get to ABI stabilty in the embedder API I guess the question would be why not do that when InitializationResult is being added to the externa API if its possible/not an unreasonable amount of work?

Reply:
> Are there other places in the embedder API where we don't go as far as CommonEnvironmentSetup?

We have numerous virtual classes, yes.

> why not do that when InitializationResult is being added to the externa API

It’s a bit easier this way, because the instance is accessed from a number of different places in our code. It doesn’t matter much, really – we can make changes while keeping ABI stability fairly easily either way.

I’ve added a small change in 7973d3de35e6928858a2dd89b599a747097532ec to make sure that there is no way to create instances of this class externally. That should rule out any theoretical concerns about userland subclasses (where virtual classes do indeed become a bit more difficult ABI-wise).

@mhdawson Please take another look.

Reply:
Thanks for the additional info.